pkg/alibabacloud/api/api.go

```diff
@@ -424,6 +429,125 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]ecs.NetworkIn
 	return result, nil
 }
 
+// describeNetworkInterfacesFromInstances lists all ENIs matching filtered ECS instances.
+// Due to a limitation in the DescribeInstances API, we can only retrieve up to 1,000 instances
+// when filtering by tags directly. To overcome this limitation, an alternative approach is
+// implemented in 3 steps:
+// 1. Filter out matching instance ids with ListTagResources
+// 2. Split instance ids into batches of 100 and send parallel DescribeInstances requests
+// 3. Split eni ids from the instances and send parallel DescribeNetworkInterfaces requests
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-listtagresources
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-describeinstances
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-describenetworkinterfaces
+func (c *Client) describeNetworkInterfacesFromInstances(ctx context.Context) ([]ecs.NetworkInterfaceSet, error) {
+	// Get filtered instance IDs
+	instanceIds := []string{}
+	instanceIdMap := make(map[string]struct{})
+	tagResouces, err := c.EcsListTagResources(ctx, c.instancesFilters)
+	if err != nil {
+		return nil, err
+	}
+
+	// The response of ListTagResources can have duplicate instanceId,
+	// remove the duplicates first.
+	for _, t := range tagResouces {
+		if _, exists := instanceIdMap[t.ResourceId]; !exists {
+			instanceIdMap[t.ResourceId] = struct{}{}
+		}
+	}
+
+	for k := range instanceIdMap {
+		instanceIds = append(instanceIds, k)
+	}
+
+	// DescribeInstances and retrieve the ENI id list. DescribeInstances accepts 100 instance IDs
+	// at most, so split instanceIds into batches of 100 and send parallel requests for performance.
+	// Return error if any request fails.
+	g := new(errgroup.Group)
+	respChan := make(chan *ecs.DescribeInstancesResponse, (len(instanceIds)/100)+1)
+
+	for i := 0; i < len(instanceIds); i += 100 {
+		idx := i
+		endIdx := idx + 100
+		if len(instanceIds) < endIdx {
+			endIdx = len(instanceIds)
+		}
+		quotedIds := make([]string, endIdx-idx)
+		for i := idx; i < endIdx; i++ {
+			quotedIds[i-idx] = fmt.Sprintf(`"%s"`, instanceIds[i])
+		}
+
+		g.Go(func() error {
+			req := ecs.CreateDescribeInstancesRequest()
+			// format: ["xxx","xxx","xxx"]
+			req.InstanceIds = fmt.Sprintf("[%s]", strings.Join(quotedIds, ","))
+			req.PageSize = requests.NewInteger(100)
+			c.limiter.Limit(ctx, "DescribeInstances")
+			resp, err := c.ecsClient.DescribeInstances(req)
+			if err != nil {
+				close(respChan)
```

Comment:
I don't think that closing `respChan` in the worker goroutines is correct. We should close it after all goroutines spawned by `g.Go` in the for loop terminates, either successfully or with an error. Otherwise the channel might be closed due to an error and a still in-flight goroutine might try to send a valid response into it later, leading to a panic.

If you want to terminate early after the first error, you might try [errgroup.WithContext](https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext)